File: sdk/python/substrait_compliance/table_data.py
"""Data structures for tabular data."""

from dataclasses import dataclass
from typing import List, Any, Optional
from enum import Enum


class DataType(Enum):
    """Supported data types."""
    INTEGER = "integer"
    BIGINT = "bigint"
    DOUBLE = "double"
    VARCHAR = "varchar"
    DATE = "date"
    BOOLEAN = "boolean"
    DECIMAL = "decimal"


@dataclass
class Column:
    """Represents a table column."""
    
    name: str
    data_type: DataType
    nullable: bool = True
    
    def __str__(self) -> str:
        null_str = "NULL" if self.nullable else "NOT NULL"
        return f"{self.name} {self.data_type.value} {null_str}"


@dataclass
class TableData:
    """Represents tabular data with schema and rows."""
    
    columns: List[Column]
    rows: List[List[Any]]
    
    def row_count(self) -> int:
        """Return number of rows."""
        return len(self.rows)
    
    def column_count(self) -> int:
        """Return number of columns."""
        return len(self.columns)
    
    def get_column(self, name: str) -> Optional[Column]:
        """Get column by name."""
        for col in self.columns:
            if col.name == name:
                return col
        return None
    
    def __str__(self) -> str:
        return f"TableData({self.column_count()} cols, {self.row_count()} rows)"

File: sdk/rust/src/table_data.rs
//! Data structures for tabular data.

use serde::{Deserialize, Serialize};

/// Supported data types.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DataType {
    Integer,
    Bigint,
    Double,
    Varchar,
    Date,
    Boolean,
    Decimal,
}

/// Represents a table column.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Column {
    pub name: String,
    pub data_type: DataType,
    pub nullable: bool,
}

impl Column {
    pub fn new(name: impl Into<String>, data_type: DataType) -> Self {
        Self {
            name: name.into(),
            data_type,
            nullable: true,
        }
    }
    
    pub fn not_null(mut self) -> Self {
        self.nullable = false;
        self
    }
}

/// Represents tabular data with schema and rows.
#[derive(Debug, Clone)]
pub struct TableData {
    pub columns: Vec<Column>,
    pub rows: Vec<Vec<String>>, // Simplified - would use proper types
}

impl TableData {
    pub fn new(columns: Vec<Column>, rows: Vec<Vec<String>>) -> Self {
        Self { columns, rows }
    }
    
    pub fn row_count(&self) -> usize {
        self.rows.len()
    }
    
    pub fn column_count(&self) -> usize {
        self.columns.len()
    }
    
    pub fn get_column(&self, name: &str) -> Option<&Column> {
        self.columns.iter().find(|c| c.name == name)
    }
}
